src/Buffer.mo

```diff
@@ -1,166 +1,1754 @@
+        resize(newUpsize(count + count2)); 
+      };
+      var i = 0;
+      while (i < count2) {
+        elements[count + i] := buffer2.getOpt i;
         i += 1;
       };
-      c
+
+      count += count2;
+    };
+
+    /// Inserts `element` at `index`, shifts all elements to the right of
+    /// `index` over by one index. Traps if `index` is greater than size.
+    ///
+    /// Runtime: O(size)
+    ///
+    /// Amortized Space: O(1), Worst Case Space: O(size)
+    public func insert(index : Nat, element : X) {
+      if (index > count) {
+        Prim.trap "Buffer index out of bounds in insert"
+      };
+      let elementsSize = elements.size();
+
+      if (count + 1 > elementsSize) {
+        let elementsSize = elements.size();
+        let elements2 = Prim.Array_init<?X>(newUpsize elementsSize, null);
+        var i = 0;
+        while (i < count + 1) {
+          if (i < index) {
+            elements2[i] := elements[i];
+          } else if (i == index) {
+            elements2[i] := ?element;
+          } else {
+            elements2[i] := elements[i - 1];
+          };
+          
+          i += 1;
+        };
+        elements := elements2;
+      } else {
+        var i : Nat = count;
+        while (i > index) {
+          elements[i] := elements[i - 1];
+          i -= 1;
+        };
+        elements[index] := ?element;
+      };
+
+      count += 1;
+    };
+
+    /// Inserts `buffer2` at `index`, shifts all elements to the right of
+    /// `index` over by size2. Traps if `index` is greater than size.
+    ///
+    /// Runtime: O(size)
+    ///
+    /// Amortized Space: O(1), Worst Case Space: O(size1 + size2)
+    public func insertBuffer(index : Nat, buffer2 : Buffer<X>) {
+      if (index > count) {
+        Prim.trap "Buffer index out of bounds in insertBuffer"
+      };
+
+      let count2 = buffer2.size();
+      let elementsSize = elements.size();
+
+      // resize and insert in one go
+      if (count + count2 > elementsSize) {
+        let elements2 = Prim.Array_init<?X>(newUpsize(count + count2), null);
+        var i = 0;
+        for (element in elements.vals()) {
+          if (i == index) {
+            i += count2;
+          };
+          elements2[i] := element;
+          i += 1;
+        };
+
+        i := 0;
+        while (i < count2) {
+          elements2[i + index] := buffer2.getOpt(i);
+          i += 1;
+        };
+        elements := elements2;
+      } 
+      // just insert
+      else {
+        var i = index;
+        while (i < index + count2) {
+          if (i < count) {
+            elements[i + count2] := elements[i];
+          };
+          elements[i] := buffer2.getOpt(i - index);
+
+          i += 1;
+        }
+      };
+
+      count += count2;
+    };
+
+    /// Resets the buffer. Capacity is set to 8.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func clear() {
+      count := 0;
```

Comment:
So I think the behavior is the same between both implementations. Unless were you suggesting I do this to avoid the function call overhead?

Reply:
Right, I considered having the current `clear()` implementation as well as another `clearToCapacity()`.

Although, that can be achieved by
`clear() + reserve(initialCapacity)` or actually `clear() + reserve(any arbitrary capacity)`

which seemed convenient enough already. Not sure what others think.

The other point I considered is that since `clear()` is only doing a constant capacity array allocation (of capacity = 8), there isn't a big performance improvement of adding a `clearToCapacity()` to do the allocation in a single pass.